data/tutorials/language/0lg_00_basic_datatypes.md

```diff
@@ -0,0 +1,844 @@
+
+```ocaml
+# #show option;;
+type 'a option = None | Some of 'a
+```
+
+The predefined type `list` is polymorphic in the same sense. It is a variant with two constructors and can hold data of any type. Here is how it is defined in the standard library:
+```ocaml
+# #show list;;
+type 'a list = [] | (::) of 'a * 'a list
+```
+
+The only bit of magic here is turning constructors into symbols. This is left unexplained in this tutorial. The types `bool` and `unit` also are regular variants, with the same magic:
+```ocaml
+# #show unit;;
+type unit = ()
+
+# #show bool;;
+type bool = false | true
+```
+
+Implicitly, product types also behave as variant types. For instance, pairs can be seen as inhabitants of this type:
+```ocaml
+# type ('a, 'b) pair = Pair of 'a * 'b;;
+type ('a, 'b) pair = Pair of 'a * 'b
+```
+
+Where `(int, bool) pair` would be written `int * bool` and `Pair (42, true)` would be written `(42, true)`. From the developer's perspective, everything happens as if such a type were declared for every possible product shape. This is what allows pattern matching on products.
+
+Even integers and floats can be seen as enumerated-like variant types, with many constructors and funky syntactic sugar. This is what allows pattern matching on those types.
+
+In the end, the only type construction that does not reduce to a variant is the function arrow type. It is possible to perform pattern matching on values of any type, except functions.
+
+#### User-Defined Polymorphic
+
+Here is an example of a variant type that combines constructors with data, constructors without data, polymorphism, and recursion:
+```ocaml
+# type 'a tree =
+  | Leaf
+  | Node of 'a * 'a tree * 'a tree;;
+type 'a tree = Leaf | Node of 'a * 'a tree * 'a tree
+```
+
+It can be used to represent arbitrarily labelled binary trees. Assuming such a tree would be labelled with integers, here is a possible way to compute the sum of the integers it contains, using recursion and pattern-matching.
+```ocaml
+# let rec sum = function
+  | Leaf -> 0
+  | Node (x, lft, rht) -> x + sum lft + sum rht;;
+val sum : int tree -> int = <fun>
+```
+
+Here is how the map function can be defined in this type:
+```ocaml
+# let rec map f = function
+  | Leaf -> Leaf
+  | Node (x, lft, rht) -> Node (f x, map f lft, map f rht);;
+val map : ('a -> 'b) -> 'a tree -> 'b tree = <fun>
+```
+
+In the OCaml community as well as in the larger functional programming community, the word *polymorphism* is used loosely. It is applied to things working in a similar fashion with various types. Several features of OCaml are polymorphic, in the broad sense, each is using a particular form of it, including its own name. In summary, OCaml has several forms of polymorphism. In most cases, the distinction between those concepts is blurred, but it sometimes needs to distinguish them. Here are the terms applicable to datatypes.
```

Comment:
```suggestion
In the OCaml community as well as in the larger functional programming community, the word *polymorphism* is used loosely. It is applied to things working in a similar fashion with various types. Several features of OCaml are polymorphic, in the broad sense. Each uses a particular form of it, including its own name. In summary, OCaml has several forms of polymorphism. In most cases, the distinction between those concepts is blurred, but it sometimes needs to distinguish them. Here are the terms applicable to datatypes:
```

When you say "each is using a particular form of it..." -- each what? Each feature of OCaml?